### PR TITLE
Use system environment for gfal2 stageout

### PR DESCRIPTION
**Pull request by juztas**
Pull request in CRAB3 to save environment: https://github.com/dmwm/CRABServer/pull/5193

Why this is needed? Discussions starting from: https://github.com/dmwm/CRABServer/issues/5186#issuecomment-206248679 and https://hypernews.cern.ch/HyperNews/CMS/get/comp-ops/2863/1.html and https://hypernews.cern.ch/HyperNews/CMS/get/computing-tools/966.html

@belforte

### CODE REVIEW

**Comment by ticoann** [comment]
@juztas, Justas, this is tested, right?

**Comment by juztas** [comment]
crab3 yes, prod no. But in the end, crab3 and prod is using the same gfal2. A bit more discussions are here: https://github.com/dmwm/CRABServer/pull/5193

**Comment by amaltaro** [comment]
We don't need to merge it for testbed, leave it open and I'll test it this week.

**Comment by ticoann** [comment]
Alan, don't worry about it. I am testing this now.

**Comment by ticoann** [comment]
Hi Justas, 
We are getting following error on the test. Do you know what could be the problem. It might not be related to your patch.

StageOutError
Message: 2016-04-25T11:02:57 : Command exited non-zero
        ErrorCode : 60311
        ModuleName : WMCore.Storage.StageOutError
        MethodName : __init__
        ErrorType : GeneralStageOutFailure
        ClassInstance : None
        FileName : /var/lib/condor/execute/dir_955045/glide_BSiHmp/execute/dir_960710/job/WMCore.zip/WMCore/Storage/StageOutError.py
        ClassName : None
        Command : #!/bin/bash
env -i X509_USER_PROXY=$X509_USER_PROXY bash -c '. $PWD/startup_environment.sh; printenv; date; gfal-copy -vvv -t 2400 -T 2400 -p   file:///var/lib/condor/execute/dir_955045/glide_BSiHmp/execute/dir_960710/job/WMTaskSpace/logArch1/logArchive.tar.gz srm://cmseos.fnal.gov:8443/srm/v2/server?SFN=/lustre/unmerged/logs/prod/2016/4/25/mohapatr_MonteCarlo_eff_T3USOSG_test2_160422_221238_3093/Production/0000/0/3cc26164-0af8-11e6-880b-00221959e7c0-0-0-logArchive.tar.gz'
            EXIT_STATUS=$?
            echo "gfal-copy exit status: $EXIT_STATUS"
            if [[ $EXIT_STATUS != 0 ]]; then
               echo "Non-zero gfal-copy Exit status!!!"
               echo "Cleaning up failed file:"
                env -i X509_USER_PROXY=$X509_USER_PROXY bash -c '. $PWD/startup_environment.sh; printenv; date; gfal-rm -vvv -t 600 srm://cmseos.fnal.gov:8443/srm/v2/server?SFN=/lustre/unmerged/logs/prod/2016/4/25/mohapatr_MonteCarlo_eff_T3USOSG_test2_160422_221238_3093/Production/0000/0/3cc26164-0af8-11e6-880b-00221959e7c0-0-0-logArchive.tar.gz '
               exit 60311
            fi
            exit 0

        LineNumber : 32
        ErrorNr : 0
        ExitCode : 151

**Comment by juztas** [comment]
prod logs not always easiest to debug. Can you provide me full logs tarball? somewhere in lxplus public?

**Comment by ticoann** [comment]
Hi Justas, I can't access the WN. It is running in OSG site.
Here is another line int the log. I will send link to your email for full condor log file.

```
bash: /var/lib/condor/execute/dir_955045/glide_BSiHmp/execute/dir_960710/job/WMTaskSpace/stageOut1/startup_environment.sh: No such file or directory
bash: gfal-copy: command not found
```

**Comment by amaltaro** [comment]
Seangchan, don't spend time on it please. I can test it tomorrow or on Wed.

**Comment by ticoann** [comment]
Hi Alan, Actually, I am not spending time on this. Actually, Ajit is testing OSG with gfal2. So I had to set it up for him anyway, that is why we are running the test. :smile:

File: src/python/WMCore/Storage/Backends/GFAL2Impl.py [hunk 1]
```diff
@@ -24,9 +24,9 @@ def __init__(self, stagein=False):
         # Next commands after separation are executed without env -i and this leads us with 
         # mixed environment with COMP and system python.
         # GFAL2 is not build under COMP environment and it had failures with mixed environment.
-        self.setups = "env -i X509_USER_PROXY=$X509_USER_PROXY sh -c '%s'"
-        self.removeCommand = self.setups % 'printenv; date; gfal-rm -vvv -t 600 %s '
-        self.copyCommand = self.setups % 'printenv; date; gfal-copy -vvv -t 2400 -T 2400 -p %(checksum)s %(options)s %(source)s %(destination)s'
+        self.setups = "env -i X509_USER_PROXY=$X509_USER_PROXY bash -c '%s'"
+        self.removeCommand = self.setups % '. $PWD/startup_environment.sh; printenv; date; gfal-rm -vvv -t 600 %s '
+        self.copyCommand = self.setups % '. $PWD/startup_environment.sh; printenv; date; gfal-copy -vvv -t 2400 -T 2400 -p %(checksum)s %(options)s %(source)s %(destination)s'
```
Comment (ticoann) [inline src/python/WMCore/Storage/Backends/GFAL2Impl.py]: In production test, $PWD is not the same location which startup_environment.sh file is created.
Is CRAB put the file in different location than wmagent?

```
/var/lib/condor/execute/dir_3127117/glide_COk0nT/execute/dir_3145575/job/WMTaskSpace/stageOut1/startup_environment.sh: No such file or directory

actual file is located at 
/var/lib/condor/execute/dir_3127117/glide_COk0nT/execute/dir_3145575/
```

**Comment by ticoann** [comment]
Ajit found the problem. It seems the location of startup_environment.sh is not match the location the command is looking for. Detail is described above.

**Comment by juztas** [comment]
Ok, please test this. I tested this on WN and seems to work:
```
Welcome to slot1@glidein_77875_776543170@node22-15.wn.iihe.ac.be!
Your condor job is running with pid(s) 125829.
bash-4.1$ pwd
/scratch/6244499.cream02.iihe.ac.be/glide_vmiu0J/execute/dir_125807
bash-4.1$ echo $PWD
/scratch/6244499.cream02.iihe.ac.be/glide_vmiu0J/execute/dir_125807
bash-4.1$ export JOBSTARTDIR=$PWD
bash-4.1$ cd WMTaskSpace/cmsRun/
bash-4.1$ ls
PSet.py  PSet.pyc  __init__.py  __init__.pyc
bash-4.1$ env -i PROXY=temp_proxy JOBSTARTDIR=$JOBSTARTDIR bash -c 'echo $JOBSTARTDIR; echo $PROXY'
/scratch/6244499.cream02.iihe.ac.be/glide_vmiu0J/execute/dir_125807
temp_proxy
bash-4.1$ env -i PROXY=temp_proxy bash -c 'echo $PWD; echo $PROXY'
/scratch/6244499.cream02.iihe.ac.be/glide_vmiu0J/execute/dir_125807/WMTaskSpace/cmsRun
temp_proxy
```
This means that CRAB3 also need to have a change in CRAB3 code to export this variable and also in crab3 wmcore branch.

**Comment by juztas** [comment]
P.S. I was not aware that production goes into directory and executes not from the main dir. This should solve it.
@mmascher tag you for CRAB3 change

**Comment by belforte** [comment]
maybe Brian's idea to save the env. in structure in a python object,
rather than on file, had its merits ?

**Comment by juztas** [comment]
@bbockelm comments?

**Comment by juztas** [comment]
@belforte what you don't like in this?

**Comment by juztas** [comment]
@ticoann @amaltaro comments? tests?
@jmarra13 this has to go also to crab3 branch (If Gurus accept it ;))) ), also https://github.com/dmwm/WMCore/commit/112598d4978e6b6960bdb38bf2da9b280bc36d4a which is not available in crab3 branch.

**Comment by ticoann** [comment]
Hi Justas, We have still problem. The problem seems to be different site using different gfal2 version and that case some errors related to check sum. (Actually I don't know the detail, but Ajit is running the test and he is the one investigating. :smile:) So I will hold on to this for a while until I talk to Ajit. But I can merge the crab version and create the new tag if that is needed. Thanks

**Comment by juztas** [comment]
Mh... Maybe this issue Ajit is getting? https://github.com/dmwm/WMCore/pull/6160

For crab, @mmascher ?

**Pull request merged**